### DISCUSSION TITLE
Axios, IOS 15.6 , undefined is not an object (evaluating 'error.response.status')

### DISCUSSION THREAD

**Discussion by AhmadNsour** [axios/axios#5871] [Q&A]
### Describe the issue

I am currently working on a mobile project, we are using Axios to call our backend Apis. its working fine for Android devices and IOS devices with version greater than 16. 

All IOS devices with version less than 16 is not working ex(V. 15.6) 

The same application is working fine on Dev and QA environments and the issue is only for the release version.

Below are the error object : 

error.response is undefined
error.request is undefined
error.config is undefined
error.message  undefined is not an object (evaluating 'error.response.status') 

### Example Code

```js
function httpGet(
    instance,
    abortController,
    url,
    headers,
    parameters,
    timeout,
    onSuccess,
    onFailure
  ) {
    var config = {
      params: parameters,
      headers: headers,
      timeout: timeout * 1,
    };

    if (abortController != null) {
      config.signal = abortController.signal;
    }

    var response = {};

    instance
      .get(url, config)
      .then(function (res) {
        response = handleHttpResponse(res);
        onSuccess(response);
      })
      .catch(function (err) {
        response = handleHttpError(err);
        onFailure(response);
      });

    return response;
  }
```

### Expected behavior

Axios should work with all IOS versions.

### Axios Version

1.3.4

### Adapter Version

http

### OS

IOS 15.6

**Comment by WillianAgostini** [comment]
Hi @AhmadNsour 

There must be some information inside the "error" variable or in the device logs.
It seems to me that the device is blocking requests for some security reasons. Can you check if you find anything in the logs and pass it to us?

**Comment by AhmadNsour** [comment]
Hi @WillianAgostini , Thank you for your comment. 

Error variable only has undefined is not an object (evaluating 'error.response.status') inside the error.message

its a release version, I don't think we can get logs from the application.

**Comment by AhmadNsour** [comment]
@WillianAgostini hello mate.

Can you please tell me what are the possible security reasons that might block the request from an IOS device?

**Comment by WillianAgostini** [comment]
I don't know if this is your case, so it would be good to have some kind of log so we can better understand the situation, but here is an example of a type of blocking that can occur.
https://stackoverflow.com/a/49374981

**Comment by AhmadNsour** [comment]
@WillianAgostini 

Thank you for your support.

If you know how to extract or get the application logs when its a release version it will help us and i will share them. 

We are not able to debug or check what is happening because it is a release version. 

Another weird thing;the application is working fine for all IOS devices with version greater than 16! And for all android devices.


The only error we are getting is the one I mentioned above.


I will check if the link you shared will be helpful and i will let you know.

Thank you

**Comment by robsterpsz** [comment]
This error does not seem to be constrained to some specific OS.
We are getting this error from a NodeJS instance running on a Linux-based container. 
```
TypeError: Cannot read properties of undefined (reading 'status')
```

**Comment by AhmadNsour** [comment]
@robsterpsz okay, did you find why it's happening and how to fix it?

**Comment by robsterpsz** [comment]
@AhmadNsour not yet. It's likely a local thing because this error is being triggered around a specific service only while we have many others also using Axios that do not experience this error. 

Either way, I wonder why these sorts of things have to fail since `error.response` could be defined by default as an empty object, and this could be stated as the default outcome when no proper response is obtained due to whatever reason.

**Comment by AhmadNsour** [comment]
@robsterpsz  our case is happening on a production app and we need to fix it asap because in order to solve it we need to change the current implementation (AXIOS) to another library and this is not an option for us.

Let us wait until someone might face it and he/she knows how to fix it or Axios themself they will come up with something.

**Comment by AhmadNsour** [comment]
@DigitalBrainJS Can you help on this?

**Comment by AhmadNsour** [comment]
@jasonsaayman Can you help on this?

**Comment by lhphong09t2** [comment]
I also got this issue on iOS Safari and the solution is window.location.reload(). How about you?

**Comment by jasonsaayman** [comment]
The error message is probably coming from your own error handler, not from Axios itself:

```txt
undefined is not an object (evaluating 'error.response.status')
```

That usually means some code inside `handleHttpError(err)` is doing this:

```js
error.response.status
```

But `error.response` is not always present.

Axios only sets `error.response` when the server returned an HTTP response, for example, `400`, `401`, `500`, etc.

If the request fails before an HTTP response is received, `error.response` will be `undefined`. Common causes are:

- Network failure
- Timeout
- Request cancellation
- CORS/browser security failure
- iOS App Transport Security is blocking the request
- TLS/certificate issue
- Invalid/unsupported certificate chain on older iOS versions
- The app is offline
- DNS/connectivity failure

So your error handling must handle all three cases:

```js
function handleHttpError(error) {
  if (axios.isAxiosError(error)) {
    if (error.response) {
      return {
        type: 'http',
        status: error.response.status,
        data: error.response.data
      };
    }

    if (error.request) {
      return {
        type: 'network',
        message: error.message,
        code: error.code
      };
    }

    return {
      type: 'setup',
      message: error.message
    };
  }

  return {
    type: 'unknown',
    message: error?.message || String(error)
  };
}
```

Your current code should also log the original Axios error before passing it into `handleHttpError`, otherwise the real error may be hidden by the secondary `error.response.status` crash:

```js
instance
  .get(url, config)
  .then(function (res) {
    response = handleHttpResponse(res);
    onSuccess(response);
  })
  .catch(function (err) {
    console.log('Axios error:', {
      message: err?.message,
      code: err?.code,
      responseStatus: err?.response?.status,
      responseData: err?.response?.data,
      hasRequest: !!err?.request,
      config: err?.config
    });

    response = handleHttpError(err);
    onFailure(response);
  });
```

Also, this function currently returns before the request finishes:

```js
var response = {};

instance.get(...).then(...).catch(...);

return response;
```

That `return response` will always return the initial empty object because Axios is asynchronous. Prefer returning the promise:

```js
function httpGet(
  instance,
  abortController,
  url,
  headers,
  parameters,
  timeout,
  onSuccess,
  onFailure
) {
  const config = {
    params: parameters,
    headers,
    timeout: timeout * 1
  };

  if (abortController != null) {
    config.signal = abortController.signal;
  }

  return instance
    .get(url, config)
    .then(function (res) {
      const response = handleHttpResponse(res);
      onSuccess(response);
      return response;
    })
    .catch(function (err) {
      const response = handleHttpError(err);
      onFailure(response);
      return response;
    });
}
```

For the iOS 15.6 part, I would specifically check TLS/certificate compatibility and App Transport Security. Older iOS versions can reject certificates/chains that newer iOS versions accept. Verify the backend certificate chain with a tool like SSL Labs, and check whether the production app has different ATS/network settings than Dev/QA.

The main fix in code is: never assume `error.response` exists. If there is no HTTP response, `error.response.status` cannot be read.
